cloudsync/tests/test_sync.py

```diff
@@ -1092,6 +1093,58 @@ def test_path_conflict_deleted_remotely(sync):
     assert not sync.providers[LOCAL].info_oid(new_local_oid)
     assert not sync.providers[REMOTE].info_oid(new_remote_oid)
 
+def test_contrived_sync_stuck_scenario(sync):
+    local_parent = "/local"
+    local_sub = "/local/sub"
+    local_file = "/local/sub/file"
+
+    remote_sub = "/remote/sub"
+    remote_file = "/remote/sub/file"
+
+    lfil = None
+
+    # local provider hash func generates event
+    def custom_hash(a):
+        if lfil:
+            sync.create_event(LOCAL, FILE, path=local_file, oid=lfil.oid, hash=lfil.hash)
+        return md5(a).digest()
+    sync.providers[LOCAL]._hash_func = custom_hash
+
+    # create local folders + file
+    sync.providers[LOCAL].mkdir(local_parent)
+    lsub_oid = sync.providers[LOCAL].mkdir(local_sub)
+    lfil = sync.providers[LOCAL].create(local_file, BytesIO(b"hello"))
+
+    # sync local file to remote
+    sync.create_event(LOCAL, FILE, path=local_file, oid=lfil.oid, hash=lfil.hash)
+    sync.run_until_clean(timeout=1)
+    log.info("TABLE 0:\n%s", sync.state.pretty_print())
+
+    # delete remotes
+    rfil_oid = sync.providers[REMOTE].info_path(remote_file).oid
+    del sync.providers[REMOTE]._fs_by_oid[rfil_oid]
+    rsub_oid = sync.providers[REMOTE].info_path(remote_sub).oid
+    del sync.providers[REMOTE]._fs_by_oid[rsub_oid]
+
+    # /sub/file entry
+    sync_entry = sync.state.lookup_oid(LOCAL, lfil.oid)
+    sync_entry._priority = 1
+    sync_entry[REMOTE].exists = TRASHED
+    sync_entry[LOCAL].sync_path = "/local/sub\\file"
+    sync.create_event(LOCAL, FILE, path=local_file, oid=lfil.oid, hash=lfil.hash)
+    sync_entry[REMOTE].changed = sync_entry[LOCAL].changed + 10
+
+    # /sub entry
+    sync_entry = sync.state.lookup_oid(LOCAL, lsub_oid)
+    sync_entry._priority = 1
+    sync.create_event(REMOTE, DIRECTORY, oid=rsub_oid, exists=TRASHED)
+
+    log.info("TABLE 1:\n%s", sync.state.pretty_print())
+    sync.run_until_clean(timeout=1) # times out when we get into the rename codepath
+    log.info("TABLE 2:\n%s", sync.state.pretty_print())
+
```

Comment:
might want to check actual end state a bit, like "files got synced"

Reply:
added check for is_discarded on all sync entries